### PR TITLE
feat: The "character_length" function handle "Binary" type

### LINKED ISSUE #7344: Internal error: The "character_length" function can only accept strings
### Describe the bug

On running the query below on the Clickbench multi file dataset, 

```sql
SELECT "CounterID", AVG(length("URL")) AS l, COUNT(*) AS c FROM hits WHERE "URL" <> '' GROUP BY "CounterID" HAVING COUNT(*) > 100000 ORDER BY l DESC LIMIT 25;
```

we get this error,

```bash
Internal error: The "character_length" function can only accept strings.. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### To Reproduce

Download the data using,

```bash
 ./benchmarks/bench.sh data clickbench_partitioned
```

A `hits_multi` directory with the parquet files will be created.

Execute the above queries,

```bash
datafusion-cli -c "CREATE EXTERNAL TABLE hits STORED AS PARQUET LOCATION 'hits_multi';" "{query}"
```

### Expected behavior

The queries should run successfully without erroring.

### Additional context

Datafusion 29.0.0

### PR DESCRIPTION
**Pull request by parkma99**
## Which issue does this PR close?

Closes #7344

### CODE REVIEW

**Comment by JayjeetAtGithub** [comment]
Thanks for working on this @parkma99. The fix works great !

File: datafusion/physical-expr/src/functions.rs [hunk 1]
```diff
@@ -545,6 +545,34 @@ pub fn create_physical_fun(
                     );
                     make_scalar_function(func)(args)
                 }
+                DataType::Binary => {
+                    let args: Vec<ColumnarValue> = args
+                        .iter()
+                        .map(|col_value| {
+                            cast_column(col_value, &DataType::Utf8, None).unwrap()
```
Comment (parkma99) [inline datafusion/physical-expr/src/functions.rs:552]: In there ,I do not know how to handle this `unwrap`.

**Comment by tustvold** [comment]
Perhaps we could implement this as part of the coercion rules as opposed to internal to the evaluation logic? See coerce_arguments_for_fun perhaps?

**Comment by parkma99** [comment]
> Perhaps we could implement this as part of the coercion rules as opposed to internal to the evaluation logic? See coerce_arguments_for_fun perhaps?

Thank you, it looks good, I will have a try lately

**Comment by parkma99** [comment]
in https://github.com/apache/arrow-datafusion/issues/7344#issuecomment-1689604037 
> But I think it might also be necessary to make the LENGTH function support binaries.

I think it is a good solution. What do you think @tustvold ?

**Comment by tustvold** [comment]
Responded on the linked ticket

**Comment by parkma99** [comment]
cc @tustvold @JayjeetAtGithub

**Pull request closed without merging**